### PR TITLE
feat(tool-search): default to the native tool_search protocol on known OpenAI/Codex hosts

### LINKED ISSUE #286: perf(tool-search): the #43 shim grows the tools array and invalidates the prompt-cache prefix
## Context

The GPT-5.6 efficiency post (<https://openai.com/index/gpt-5-6-frontier-intelligence-efficiency/>)
spells out the design rule behind Codex's prompt-cache hit rate:

> To preserve that prefix, the harness treats all model-visible history as append-only: new messages,
> tool results, and environment updates are added at the end rather than inserted into earlier
> context. Tools are also presented in a deterministic order […]

This is exactly why codex's native tool-search protocol delivers a revealed tool inside a
`tool_search_output` item at the **tail** of `input`, rather than by adding it to the `tools` array —
`tools` is part of the cacheable prompt prefix, `input`'s tail is not.

## Problem

shunt implements progressive tool reveal two ways, and only one of them preserves the prefix.

**Native path (#82) — correct.** Every deferred tool is withheld from `tools` unconditionally; loaded
ones ride in a `tool_search_output` (`src/model/responses_request.rs:505`). Append-only, prefix
intact.

**Text shim (#43) — invalidates the prefix.** `src/model/responses_request.rs:782-797`:

```rust
.filter(|tool| {
    let name = tool.get("name").and_then(Value::as_str).unwrap_or("");
    if !context.deferred_names.contains(name) {
        return true;
    }
    !context.native && context.loaded_tools.contains(name)   // <-- shim: tool joins `tools`
})
```

Once a `tool_reference` loads a deferred tool, that tool is added to the `tools` array and stays there
for the rest of the session. The array is re-derived per turn from the client's history, so the
resulting Responses request has a `tools` array that *grows* mid-conversation. Each growth changes the
prompt prefix, so the cached prefix is dropped and the whole conversation is re-processed from the
tool boundary onward.

The cost lands worst exactly where tool search is supposed to help: long sessions with many deferred
tools reveal repeatedly, and each reveal costs a full prefix recompute.

## The shim is the default

`src/config.rs:2796` requires three conditions for the native path, and the provider opt-in
(`tool_search = true`) is **off by default** — so the shim is what most users get today, including on
the `Chatgpt` and `OpenAi` flavors where the native protocol is known to work.

## Proposed

1. Flip `tool_search` to default-on for the `Chatgpt` / `OpenAi` flavors, where `native_tool_search`
   already gates on flavor and model support. xAI/Grok keep the shim regardless.
2. If (1) is considered too aggressive to default, document the cache cost of the shim in the
   tool-search reference page so the trade-off is a deliberate choice.

Either way the shim must stay for the flavors that cannot do native tool search — this is about which
path is the default, not about removing one.

## Verification

Cross-checked against `openai/codex` @ `6219b7c`. Also confirmed already-correct and needing no change:
tool ordering is preserved from the client array (deterministic), `prompt_cache_key` is stable per
session (`src/model/responses_request.rs:199-239`), account selection is session-sticky
(`src/accounts.rs:950`), and the WS path already sends `previous_response_id` + input delta
(`src/adapters/responses/codex_continuation.rs`) — the post's "incremental transport".

A before/after measurement on `input_tokens_details.cached_tokens` across a session with at least one
reveal would quantify the win; that counter is already observed in `src/stream_metrics.rs:406`.

## Docs

`site/src/content/docs/reference/` — the tool-search config key default and its cache implications.

### PR DESCRIPTION
**Pull request by amondnet**
## Summary

`tool_search` now defaults to `true`: shunt implements Claude Code's progressive tool reveal two ways, and only one preserves the prompt-cache prefix. The #43 text shim adds each revealed deferred tool to the request's `tools` array (`src/model/responses_request.rs:791-797`), and `tools` is part of the cacheable prompt prefix shunt re-derives from client history every turn — so each reveal grows the array, invalidates the cached prefix, and makes the backend re-process the conversation from the tool boundary onward. The cost lands worst in exactly the long, many-reveal sessions tool search is meant to help. The native `tool_search` protocol (#82) withholds every deferred tool from `tools` and delivers a revealed one in a `tool_search_output` item appended at the tail of `input` — the append-only-history rule OpenAI describes as the source of Codex's prompt-cache hit rate.

**Worse than the issue described.** On reveal the shim ALSO injects the revealed tool's full schema as history text (`tool_reference_text`, `src/model/responses_request.rs:623-643`). So from that point on the schema ships twice per turn — once as history text, once in `tools` — in addition to killing the cached prefix. This double-send is unavoidable without the native protocol, because the Responses API only lets the model call a tool that appears in `tools`.

`Config::native_tool_search`'s gating logic is unchanged, so the existing gates still decide: the flavor must be stock OpenAI or ChatGPT/Codex and the routed model gpt-5.4+. xAI/Grok routes and gpt-5.2-and-below keep the shim, silently, as before. `tool_search = false` is the per-provider opt-out. The shim itself is not modified — it stays for the flavors that cannot do native tool search.

**The default had to change in two places.** `#[serde(default = "default_true")]` governs user-declared providers and `[[upstreams]]` entries, while the struct literals in `Config::default()` govern the built-in provider names — because `Config::load` seeds figment with `Serialized::defaults(Config::default())`. Flipping only the serde attribute would have left `codex`/`openai` on the shim, and no unit test that builds `Config` directly in Rust would have caught it. Hence new coverage at the `Config::load` boundary for both declaration forms (`[[upstreams]]` needs its own, since `normalize()` builds its `ProviderConfig` independently).

**Risk.** Narrow: the native path only engages when the client actually sends `defer_loading` tools, i.e. under `ENABLE_TOOL_SEARCH=true`, which Claude Code keeps off behind a non-first-party base URL unless the user opts in. The native shapes were live-probe verified against the ChatGPT/Codex backend on gpt-5.6 (2026-07-13), and `tool_search` is documented on the public OpenAI Responses API for gpt-5.4+ models — though api.openai.com has not been probed by us directly.

**Not done.** No before/after measurement of `input_tokens_details.cached_tokens` across a session with a reveal (the issue mentions it as a way to quantify the win); that needs a real Codex session and was deliberately skipped. The cache argument here rests on the wire shapes plus the prefix-stability test.

## Milestone / spec

#43 (text shim), #82 (native `tool_search` protocol), #286 (this change)

## Tests (7 added/renamed)

- `native_tool_search_requires_opt_in_flavor_and_model` reworked into `native_tool_search_defaults_on_and_gates_on_flavor_and_model`, keeping every still-valid assertion (xai blocked, unknown provider, the `gpt-5.40`/`gpt-5.41-turbo` boundary cases, `gpt-5.2-codex`)
- `native_tool_search_false_forces_shim_opt_out`
- Load-boundary default + opt-out tests for `[providers.*]` and for `[[upstreams]]`
- `tool_reveal_grows_shim_tools_but_leaves_native_tools_stable`, which pins the mechanic directly — shim `tools` grows across a reveal, native `tools` is byte-identical pre/post with the revealed schema riding in the appended `tool_search_output`

## Docs

The new default is documented across en/ko/ja/zh-cn (`reference/configuration.md`, `guides/codex.mdx`, `reference/troubleshooting.md`, `getting-started/comparison.mdx`) plus `README.md`, `docs/comparison.md`, and `providers/openai.mdx`. The `Native protocol (opt-in)` heading was renamed and every per-locale anchor reference updated. Separately, four places still described pre-#43 shim behavior — claiming it "forwards all deferred schemas upfront" and "reclaims zero context", and a footnote claiming shunt renders a name-only `"Loaded tool: X"` when that is only the unknown-reference fallback — all corrected with accurate line citations.

## Checklist

- [x] `cargo build` passes
- [x] `cargo test` passes (new behavior is covered; tests run without network/loopback where possible)
- [x] `cargo clippy --all-targets -- -D warnings` clean
- [x] `cargo fmt --all --check` clean
- [x] Source files stay under 500 lines
- [x] English only; matches surrounding style
- [x] Frozen spec in `docs/` updated if this change deviates from it
- [x] User-facing docs updated for behavior/config/endpoint/CLI/provider/model changes — `README.md` / `site/` as applicable (`wiki/` is generated; don't hand-edit)
- [ ] Any new GitHub Action is pinned to a full commit SHA (n/a — no workflow changes)

`cargo fmt --all --check`, `cargo clippy --all-targets --all-features -- -D warnings`, and `cargo test --all-features --workspace` all pass (exit 0), re-run against the final diff.

## Notes for reviewers

- The double-send point above (`tool_reference_text` injecting schema as history text on top of the growing `tools` array) is the sharpest edge — worth confirming against `src/model/responses_request.rs:623-643` and `:791-797`.
- The two-places default change (`serde` attribute vs. `Config::default()` struct literals, plus `[[upstreams]]`'s independent `normalize()` path) is easy to half-fix; the added load-boundary tests are meant to guard exactly that.
- No cache-hit measurement was taken; the native-path win is argued from wire shapes and the prefix-stability test only.

Closes #286

---
## Summary by cubic
Default to the Responses API’s native `tool_search` on GPT‑5.4+ for stock OpenAI/Codex, but only on hosts known to implement it (ChatGPT/Codex and `api.openai.com`); all other OpenAI‑compatible endpoints keep the #43 text shim unless opted in. This preserves the prompt‑cache prefix and avoids double‑sending tool schemas. Addresses #286.

- **New Features**
  - `tool_search` is now tri‑state: unset (“auto”) uses native only on known hosts; `true` forces native; `false` forces the shim. Gate checks provider flag, flavor, model, and host; xAI/Grok and ≤GPT‑5.2 stay on the shim.
  - Native path keeps `tools` stable and emits reveals as `tool_search_output` in `input`; adds a debug log of the resolved protocol with no silent fallback.

- **Migration**
  - To force the shim, set `tool_search = false` under `[providers.<name>]` or export `SHUNT_PROVIDERS__<NAME>__TOOL_SEARCH=false`.
  - To use native on a custom OpenAI‑compatible endpoint, set `tool_search = true` and route to a GPT‑5.4+ model on a stock OpenAI/Codex flavor.

<sup>Written for commit aa18e45fe1e0318ec93c6033c6ae317dd15361ec. Summary will update on new commits.</sup>

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -1133,17 +1133,19 @@ pub struct ProviderConfig {
...
+    /// prompt prefix, so every reveal invalidates the cached prefix and forces
+    /// a full re-process; native keeps deferred tools out of `tools` and
+    /// appends a `tool_search_output` item instead, leaving the prefix stable
+    /// (issue #286). Only honored where the upstream flavor and model support
+    /// it (see [`Config::native_tool_search`]) — stock OpenAI and ChatGPT-Codex
+    /// on gpt-5.4+ take the native path; xAI/Grok and older models keep the
+    /// shim regardless. Set `tool_search = false` to force the shim back on
+    /// for a provider that does support it.
+    #[serde(default = "default_true")]
     pub tool_search: bool,
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/config.rs]: [claude] Confirmed and fixed in aa18e45 — thank you, this was correct and the framing was the useful part: `OpenAi` is the catch-all `else` in `responses_flavor`, not "stock OpenAI", so the default-on flip did reach arbitrary OpenAI-compatible endpoints.

`tool_search` is now tri-state:

- `Some(true)` — force native (still gated on flavor + model).
- `Some(false)` — force the #43 shim.
- unset (`None`, "auto", the default) — native **only** for an upstream already known to accept the protocol: `AuthMode::ChatgptOauth` (the ChatGPT/Codex backend) or a `base_url` whose host is exactly `api.openai.com`.

LiteLLM, vLLM, OpenRouter and self-hosted proxies therefore keep the shim until their operator opts in explicitly, which is the behavior you asked for.

The host match is exact rather than suffix-based on purpose — this is about trusting one specific `tool_search` implementation, and `chat.openai.com` / `platform.openai.com` are different products.

Covered by four tests spanning the quadrants (`src/config.rs`): `legacy_provider_custom_host_tool_search_defaults_to_shim`, `..._true_opts_into_native`, `legacy_provider_on_known_openai_host_defaults_on` (asserts the rule reads the host, not the provider name), and `legacy_provider_tool_search_false_opts_out`.
